spvenvelope.go

```diff
@@ -0,0 +1,179 @@
+}
+
+// MapiResponse is a callback from mApi
+type MapiResponse struct {
+	CallbackPayload MerkleProof `json:"callbackPayload"`
+	APIVersion      string      `json:"apiVersion"`
+	Timestamp       time.Time   `json:"timestamp"`
+	MinerID         string      `json:"minerId"`
+	BlockHash       string      `json:"blockHash"`
+	BlockHeight     uint64      `json:"blockHeight"`
+	CallbackTxID    string      `json:"callbackTxId"`
+	CallbackReason  string      `json:"callbackReason"`
+}
+
+// VerifyPayment verifies whether or not the txs supplied via the supplied SPVEnvelope are valid
+func (s *SPVClient) VerifyPayment(ctx context.Context, payment *SPVEnvelope) (bool, error) {
+	proofs := make(map[string]bool)
+	outputValues := make(map[string]uint64)
+
+	tx, err := bt.NewTxFromString(payment.RawTX)
+	if err != nil {
+		return false, err
+	}
+
+	valid, err := s.verifyTxs(ctx, payment, nil, true, proofs)
+	if err != nil {
+		return false, err
+	}
+	if !valid {
+		return valid, nil
+	}
+
+	for _, v := range proofs {
+		if !v {
+			return false, ErrPaymentNotVerified
+		}
+	}
+
+	outputValues[tx.GetTxID()] = tx.GetTotalOutputSatoshis()
+
+	return true, nil
+}
+
+func (s *SPVClient) verifyTxs(ctx context.Context, payment *SPVEnvelope, childInputs []*bt.Input,
+	isRoot bool, proofs map[string]bool) (bool, error) {
+	tx, err := bt.NewTxFromString(payment.RawTX)
+	if err != nil {
+		return false, err
+	}
+	txID := tx.GetTxID()
+	proofs[txID] = false
+
+	if isRoot && payment.Proof != nil {
+		return false, ErrRootPaymentConfirmed
+	}
+
+	for inputTxID, input := range payment.Inputs {
+		if input.TxID == "" {
+			input.TxID = inputTxID
+		}
+		valid, err := s.verifyTxs(ctx, input, tx.GetInputs(), false, proofs)
+		if err != nil {
+			return false, err
+		}
+		if !valid {
+			return valid, nil
+		}
+	}
+
+	// cannot verify the proof or outputs of the root tx
+	if isRoot {
+		proofs[txID] = true
+		return true, nil
+	}
+
+	// if at the leafs of tree and transaction is unconfirmed, fail
+	if (payment.Inputs == nil || len(payment.Inputs) == 0) && payment.Proof == nil {
+		return false, ErrNoConfirmedTransaction
+	}
+
+	//childTxInputMap := make(map[string]bool)
+	if payment.Proof != nil {
+		proofTxID := payment.Proof.TxOrID
+		if len(proofTxID) != 64 {
+			proofTx, err := bt.NewTxFromString(payment.Proof.TxOrID)
+			if err != nil {
+				return false, err
+			}
+
+			proofTxID = proofTx.GetTxID()
+		}
+
+		if proofTxID != payment.TxID {
+			return false, ErrTxIDMismatch
+		}
+
+		var proofPresent bool
+		for _, childInput := range childInputs {
+			if childInput.PreviousTxID == proofTxID {
+				proofPresent = true
+				break
+			}
+		}
+
+		if !proofPresent {
+			return false, ErrProofTxMismatch
+		}
+
+		valid, _, err := s.VerifyMerkleProofJSON(ctx, payment.Proof)
+		if err != nil {
+			return false, err
+		}
+
+		proofs[txID] = valid
+
+		return valid, nil
+	}
+
+	var pass bool
+	for _, input := range childInputs {
```

Comment:
could this stuff not be in the above for loop? not exactly sure what the stuff below is doing tbh...

Reply:
This will is going to be what verifies the inputs of child tx against the outputs of its parent. It's here a placeholder while we're dev'ing the engine in go-bt. Would you rather it was just left it out?

Reply:
we shouldn't need to verify the scripts for txs that are already in the blockchain - we assume miners already verified them or else they wouldn't be mined into the blockchain (and have a merkle proof). we should only need to verify the script of the tx being sent (the root tx) - or if the input txs haven't been mined yet and don't have a mapi response on them, but let's leave that edge case for now

Reply:
also here, could we put this in a named function for clarity?